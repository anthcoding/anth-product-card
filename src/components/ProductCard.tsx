import React from 'react';
import styles from '../styles/styles.module.css'
import { createContext } from 'react';
import { useProduct } from '../hooks/useProduct';
import { CounterContextProps, Product, onChangeArgs, InitialValues, ProductCartHandlers } from '../interfaces/interfaces';

export const ProductContext = createContext({} as CounterContextProps)
const { Provider } = ProductContext;

export interface Props {
	// children: ReactElement | ReactElement[];
	children: (args: ProductCartHandlers) => JSX.Element;
	product: Product;
	className?: string;
	style?: React.CSSProperties;
	onChange?: (args: onChangeArgs) => void;
	value?: number;
	initialValues?: InitialValues;
}

export const ProductCard = ({ children, product, className, style, onChange, value, initialValues }: Props) => {

	const { counter, increaseBy, maxC, isMaxCountReached, reset } = useProduct({
		initialValues,
		onChange,
		product,
		value,
	})

	return (
		<Provider value={{
			counter,
			increaseBy,
			product,
			maxC,
		}}>
			<div className={`${styles.productCard} ${className}`}
				style={style}
			>
				{children({
					count: counter,
					isMaxCountReached,
					isMaxCount: initialValues?.maxCount,
					product,

					increaseBy,
					reset
				})}
			</div>
		</Provider>
	)
}
