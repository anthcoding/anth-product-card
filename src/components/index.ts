export {ProductButtons} from "./ProductButtons";
export {ProductCard} from './ProductCard';
export {ProductImage} from './ProductImage';
export {ProductTitle} from './ProducTitle';
