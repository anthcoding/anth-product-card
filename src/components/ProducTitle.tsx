import React from 'react';
import { useContext } from "react"
import styles from '../styles/styles.module.css'
import { ProductContext } from "./ProductCard";

interface TitleProps {
	className?: string;
	title?: string;
	style?: React.CSSProperties
}

export const ProductTitle = ({ title, style }: TitleProps) => {
	const { product } = useContext(ProductContext)

	return (
		<span className={`${styles.productDescription}`} style={style} >{title ? title : product.title}</span>
	)
}
