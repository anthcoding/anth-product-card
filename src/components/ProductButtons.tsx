import React from "react";
import { useContext } from "react";
import styles from '../styles/styles.module.css'
import { ProductContext } from "./ProductCard";

interface ButtonsProps {
	className?: string;
	style?: React.CSSProperties;
}

export const ProductButtons = ({ className, style }: ButtonsProps) => {
	const { increaseBy, counter, maxC } = useContext(ProductContext);

	console.log('Hoola')

	return (
		<div className={`${styles.buttonsContainer} ${className}`}
			style={style}
		>
			<button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>-</button>
			<div className={styles.countLabel}> {counter} </div>
			<button className={`${styles.buttonAdd} ${maxC && counter >= maxC ? 'disabled' : ''}`} onClick={() => increaseBy(1)}>+</button>
		</div>
	);
}
