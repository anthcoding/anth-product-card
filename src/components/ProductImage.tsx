import React from 'react';
import { useContext } from "react";
import noImage from '../assets/no-image.jpg'
import styles from '../styles/styles.module.css'
import { ProductContext } from "./ProductCard";

interface ImageProps {
	className?: string;
	img?: string;
	style?: React.CSSProperties
}

export const ProductImage = ({ img = '', className, style }: ImageProps) => {

	const { product } = useContext(ProductContext)

	let imgToShow: string;

	if (img) {
		imgToShow = img;
	} else if (product.img) {
		imgToShow = product.img;
	} else {
		imgToShow = noImage;
	}

	return (
		<img
			className={`${styles.productImg} ${className}`}
			src={imgToShow}
			alt="Product img"
			style={style}
		/>
	)
}