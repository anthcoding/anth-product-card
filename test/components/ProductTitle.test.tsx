import React from 'react';
import renderer from 'react-test-renderer'
import { ProductTitle, ProductCard } from '../../src/components';
import { product1 } from '../data/products';


describe('Product Title', () => {
	test("Must show the component correctly with the personalized title", () => {

		const wrapper = renderer.create(
			<ProductTitle title='Custom Product' />
		)
		expect(wrapper.toJSON()).toMatchSnapshot()

	})

	test("Must show the component with the name of the product", () => {

		const wrapper = renderer.create(
			<ProductCard product={product1}>
				{
					() => (
						<ProductTitle />
					)
				}
			</ProductCard>
		)
		expect(wrapper.toJSON()).toMatchSnapshot()

	})
});
