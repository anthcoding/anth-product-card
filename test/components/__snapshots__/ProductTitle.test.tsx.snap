// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Title Must show the component correctly with the personalized title 1`] = `
<span
  className="productDescription"
>
  Custom Product
</span>
`;

exports[`Product Title Must show the component with the name of the product 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription"
  >
    Coffee Mug - Without Img
  </span>
</div>
`;
